### PR TITLE
update Production.py to re-reun tau reco at miniaod, add extended MC truth to the ntuples

### PR DESCRIPTION
**Pull request by lguzzi**
Addressing todo's:
- Re-run tau reco on MINIAOD with updated signal cone and removed DM finding filter (#1)
- add extended mc truth information (#2)
- set cone for pf cand around jets to 0.8 (#5)

### CODE REVIEW

File: Production/python/Production.py [hunk 1]
```diff
@@ -69,16 +74,22 @@
...
+
+process.combinatoricRecoTaus.builders[0].signalConeSize = cms.string('max(min(0.2, 4.528/(pt()^0.8982)), 0.03)') ## change to quantile 0.95
+process.selectedPatTaus.cut = cms.string('pt > 18.')   ## remove DMFinding filter (was pt > 18. && tauID(\'decayModeFindingNewDMs\')> 0.5)
+
+tauIdConfig = importlib.import_module('TauMLTools.Production.runTauIdMVA')
 updatedTauName = "slimmedTausNewID"
 tauIdEmbedder = tauIdConfig.TauIDEmbedder(
     process, cms, debug = False, updatedTauName = updatedTauName,
-    toKeep = [ "2017v2", "dR0p32017v2", "newDM2017v2", "deepTau2017v2p1", "deepTau2017v2p1", "againstEle2018",  ]
+    toKeep = [ "2017v2", "dR0p32017v2", "newDM2017v2", "deepTau2017v2p1"]
```
Comment (kandrosov) [inline Production/python/Production.py:87]: If `2017v2`, `dR0p32017v2` or `newDM2017v2` can be rerun, they should be stored in the root-tuple.

File: Production/python/Production.py [hunk 2]
```diff
@@ -69,16 +74,22 @@
...
 )
-tauIdEmbedder.runTauID()
+tauIdEmbedder.runTauID(tau_collection = 'selectedPatTaus')
 tauSrc_InputTag = cms.InputTag('slimmedTausNewID')
 
-tauJetdR = 0.2
+tauJetdR = 0.8
```
Comment (kandrosov) [inline Production/python/Production.py:92]: should be 0.3: 0.8 - 0.5

File: Analysis/interface/TauTuple.h [hunk 3]
```diff
@@ -100,6 +106,14 @@
...
                                                       the tau isolation cone with dR < 0.3 */ \
     VAR(Float_t, puCorrPtSum) /* pile-up correction for the sum of the transverse momentums */ \
+    VAR(Int_t, n_charged_hadrons) \
+    VAR(Int_t, n_neutral_hadrons) \
+    VAR(Int_t, n_gammas) \
+    VAR(Int_t, n_gammas_rad) \
+    VAR4(Float_t, lepMatch_visible_pt , lepMatch_visible_eta, \
+                  lepMatch_visible_phi, lepMatch_visible_mass) \
+    VAR4(Float_t, lepMatch_visible_rad_pt , lepMatch_visible_rad_eta, \
+                  lepMatch_visible_rad_phi, lepMatch_visible_rad_mass) \
```
[the remark this answers is not in the document]
Comment (lguzzi) [reply Analysis/interface/TauTuple.h]: done, see https://github.com/cms-tau-pog/TauMLTools/pull/13/commits/4d3c780492537eed4dc1ed89e1089f2586671ec9

File: Production/python/Production.py [hunk 4]
```diff
@@ -4,6 +4,9 @@
 import importlib
 import FWCore.ParameterSet.Config as cms
 from FWCore.ParameterSet.VarParsing import VarParsing
+import RecoTauTag.Configuration.tools.adaptToRunAtMiniAOD as tauAtMiniTools
+from Configuration.Eras.Era_Run2_2018_cff import Run2_2018
```
[the remark this answers is not in the document]
Comment (lguzzi) [reply Production/python/Production.py]: done, see https://github.com/cms-tau-pog/TauMLTools/pull/13/commits/b36439fcb422845b499639b8b29e58f09d1dd9ad

File: Production/python/Production.py (hunk 2 above)
Comment (lguzzi) [reply Production/python/Production.py]: fixed, see https://github.com/cms-tau-pog/TauMLTools/pull/13/commits/eb0c5b9a8440bb74fa54d2eeef676b0d38489d9e

File: Production/python/Production.py [hunk 5]
```diff
@@ -26,15 +29,17 @@
                  "Store jets that don't match to any pat::Tau.")
 options.register('requireGenMatch', True, VarParsing.multiplicity.singleton, VarParsing.varType.bool,
                  "Store only taus/jets that have GenLeptonMatch or GenQcdMatch.")
+options.register('reclusterJets', True, VarParsing.multiplicity.singleton, VarParsing.varType.bool,
+                " If 'reclusterJets' set true a new collection of uncorrected ak4PFJets is built to seed taus (as at RECO), otherwise standard slimmedJets are used")
```
[the remark this answers is not in the document]
Comment (lguzzi) [reply Production/python/Production.py:32]: done, see https://github.com/cms-tau-pog/TauMLTools/pull/13/commits/705430133fa63822b34595f2427e43ec1d90f66a

File: Production/python/Production.py (hunk 1 above)
Comment (lguzzi) [reply Production/python/Production.py:92]: done, see https://github.com/cms-tau-pog/TauMLTools/pull/13/commits/e26149c6cd6c154593046ad5775c1e201722ec20

File: Analysis/interface/TauTuple.h [hunk 6]
```diff
@@ -68,10 +68,16 @@
     VAR(Int_t, lepton_gen_charge) /* charge of the matched gen lepton */ \
     VAR4(Float_t, lepton_gen_pt, lepton_gen_eta, \
                   lepton_gen_phi, lepton_gen_mass) /* 4-momentum of the matched gen lepton */ \
+    VAR(Int_t, lepton_gen_firstCopy_charge) \
```
[the remark this answers is not in the document]
Comment (lguzzi) [reply Analysis/interface/TauTuple.h]: done, see https://github.com/cms-tau-pog/TauMLTools/pull/13/commits/f419a7570bea700b04e9a0e6aac6bfaf43640b0f

File: Production/python/sampleConfig.py [hunk 7]
```diff
@@ -52,3 +61,10 @@ def GetGlobalTag(sampleType):
         print "ERROR: unknown sample type = '{}'".format(sampleType)
         sys.exit(1)
     return globalTagMap[sampleType]
+
+def GetPeriodCfg(sampleType):
+    period = GetPeriod(sampleType)
+    if period not in periodCfgMap:
+        print "ERROR: unknown sample type = '{}'".format(sampleType)
+        sys.exit(1)
+    return periodCfgMap[GetPeriod(sampleType)]
```
Comment (kandrosov) [inline Production/python/sampleConfig.py:70]: `GetPeriod(sampleType)` -> `period`

File: Analysis/interface/TauTuple.h [hunk 8]
```diff
@@ -65,13 +65,21 @@
     VAR(Int_t, tau_charge) /* tau charge */ \
     VAR(Int_t, lepton_gen_match) /* matching with leptons on the generator level (see Htautau Twiki for details):
                                     Electron = 1, Muon = 2, TauElectron = 3, TauMuon = 4, Tau = 5, NoMatch = 6 */\
-    VAR(Int_t, lepton_gen_charge) /* charge of the matched gen lepton */ \
+    VAR(Int_t, lepton_gen_charge) /* charge of the matched gen lepton (last copy)*/ \
     VAR4(Float_t, lepton_gen_pt, lepton_gen_eta, \
-                  lepton_gen_phi, lepton_gen_mass) /* 4-momentum of the matched gen lepton */ \
+                  lepton_gen_phi, lepton_gen_mass) /* 4-momentum of the matched gen lepton (last copy)*/ \
+    VAR(Int_t, lepton_gen_firstCopy_charge)    /* charge of the matched gen lepton (first copy) */ \
```
[the remark this answers is not in the document]
Comment (lguzzi) [reply Analysis/interface/TauTuple.h]: removed in https://github.com/cms-tau-pog/TauMLTools/pull/13/commits/8931c68bbbc26e902646b3f781ddeef46de0cf37

File: Analysis/interface/TauTuple.h [hunk 9]
```diff
@@ -100,6 +108,10 @@
     VAR(Float_t, photonPtSumOutsideSignalConedR03) /* sum of the transverse momentums of photons inside
                                                       the tau isolation cone with dR < 0.3 */ \
     VAR(Float_t, puCorrPtSum) /* pile-up correction for the sum of the transverse momentums */ \
+    VAR(Int_t, n_charged_hadrons) /* number of charged hadrons in the final state */ \
+    VAR(Int_t, n_neutral_hadrons) /* number of neutral hadrons in the final state */ \
+    VAR(Int_t, n_gammas) /* number of photons in the final state */ \
+    VAR(Int_t, n_gammas_rad) /* number of photons from the final state radiation */ \
```
[the remark this answers is not in the document]
Comment (lguzzi) [reply Analysis/interface/TauTuple.h]: removed in https://github.com/cms-tau-pog/TauMLTools/pull/13/commits/8931c68bbbc26e902646b3f781ddeef46de0cf37

File: Production/python/sampleConfig.py (hunk 7 above)
Comment (lguzzi) [reply Production/python/sampleConfig.py]: fixed https://github.com/cms-tau-pog/TauMLTools/pull/13/commits/05955de746abc3d44842ed728f87e121bdbde41d

**Comment by kandrosov** [comment]
- Production.py correctly runs with `rerunTauReco=True` and `rerunTauReco=False`
- new variables filled correctly

**Pull request merged**